File: ts/tests/batch.test.ts
import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { freshStore, toBytes, fromBytes, rmTmpDir, fs } from './helpers.js';
import {
  GitStore,
  FS,
  FileType,
  FileNotFoundError,
  IsADirectoryError,
  PermissionError,
  StaleSnapshotError,
  MODE_BLOB_EXEC,
  MODE_LINK,
} from '../src/index.js';
import * as path from 'node:path';

let store: GitStore;
let snap: FS;
let tmpDir: string;

beforeEach(async () => {
  const res = await freshStore();
  store = res.store;
  tmpDir = res.tmpDir;
  let f = await store.branches.get('main');
  snap = await f.write('a.txt', toBytes('a'));
});

afterEach(() => rmTmpDir(tmpDir));

describe('batch', () => {
  it('multiple writes single commit', async () => {
    const b = snap.batch({ message: 'bulk' });
    await b.write('x.txt', toBytes('x'));
    await b.write('y.txt', toBytes('y'));
    const f2 = await b.commit();
    expect(await f2.exists('x.txt')).toBe(true);
    expect(await f2.exists('y.txt')).toBe(true);
    // Only one commit ahead
    const parent = await f2.getParent();
    expect(parent!.commitHash).toBe(snap.commitHash);
  });

  it('custom message', async () => {
    const b = snap.batch({ message: 'bulk upload' });
    await b.write('x.txt', toBytes('x'));
    const f2 = await b.commit();
    expect(await f2.getMessage()).toBe('bulk upload');
  });

  it('write and remove', async () => {
    const b = snap.batch({ message: 'mixed' });
    await b.write('new.txt', toBytes('new'));
    await b.remove('a.txt');
    const f2 = await b.commit();
    expect(await f2.exists('new.txt')).toBe(true);
    expect(await f2.exists('a.txt')).toBe(false);
  });

  it('empty batch returns same FS', async () => {
    const b = snap.batch({ message: 'empty' });
    const f2 = await b.commit();
    expect(f2.commitHash).toBe(snap.commitHash);
  });

  it('batch on readonly raises', async () => {
    const f1 = await snap.write('x.txt', toBytes('x'));
    await store.tags.set('v1', f1);
    const tagged = await store.tags.get('v1');
    expect(() => tagged.batch()).toThrow(/read-only/);
  });

  it('last op wins: write then remove', async () => {
    const b = snap.batch();
    await b.write('new.txt', toBytes('data'));
    await b.remove('new.txt');
    // new.txt was only staged, not in base, so remove should un-stage it
    // and since it's not in base, removing should throw
    // Actually: write stages it, remove checks if it exists in base (no) and was staged (yes)
    // Looking at the code: remove deletes from _writes and only adds to _removes if existsInBase
    const f2 = await b.commit();
    expect(await f2.exists('new.txt')).toBe(false);
  });

  it('overwrite then remove existing', async () => {
    const b = snap.batch();
    await b.write('a.txt', toBytes('updated'));
    await b.remove('a.txt');
    const f2 = await b.commit();
    expect(await f2.exists('a.txt')).toBe(false);
  });

  it('remove then write', async () => {
    const b = snap.batch();
    await b.remove('a.txt');
    await b.write('a.txt', toBytes('new content'));
    const f2 = await b.commit();
    expect(fromBytes(await f2.read('a.txt'))).toBe('new content');
  });

  it('write after commit raises', async () => {
    const b = snap.batch();
    await b.write('x.txt', toBytes('x'));
    await b.commit();
    await expect(b.write('y.txt', toBytes('y'))).rejects.toThrow(/closed/);
  });

  it('remove after commit raises', async () => {
    const b = snap.batch();
    await b.commit();
    await expect(b.remove('a.txt')).rejects.toThrow(/closed/);
  });

  it('remove directory raises IsADirectoryError', async () => {
    const f2 = await snap.write('d/f.txt', toBytes('x'));
    const b = f2.batch();
    await expect(b.remove('d')).rejects.toThrow(IsADirectoryError);
  });

  it('remove missing raises FileNotFoundError', async () => {
    const b = snap.batch();
    await expect(b.remove('nope.txt')).rejects.toThrow(FileNotFoundError);
  });

  it('stale batch is retryable', async () => {
    const stale = snap;
    await snap.write('x.txt', toBytes('x')); // advance branch
    const b = stale.batch();
    await b.write('y.txt', toBytes('y'));
    await expect(b.commit()).rejects.toThrow(StaleSnapshotError);
  });

  it('writeFromFile basic', async () => {
    const filePath = path.join(tmpDir, 'local.txt');
    fs.writeFileSync(filePath, 'from disk');
    const b = snap.batch();
    await b.writeFromFile('repo.txt', filePath);
    const f2 = await b.commit();
    expect(await f2.readText('repo.txt')).toBe('from disk');
  });

  it('writeFromFile preserves executable', async () => {
    const filePath = path.join(tmpDir, 'run.sh');
    fs.writeFileSync(filePath, '#!/bin/sh');
    fs.chmodSync(filePath, 0o755);
    const b = snap.batch();
    await b.writeFromFile('run.sh', filePath);
    const f2 = await b.commit();
    expect(await f2.fileType('run.sh')).toBe(FileType.EXECUTABLE);
  });

  it('writeFromFile mode override', async () => {
    const filePath = path.join(tmpDir, 'f.txt');
    fs.writeFileSync(filePath, 'data');
    const b = snap.batch();
    await b.writeFromFile('f.txt', filePath, { mode: MODE_BLOB_EXEC });
    const f2 = await b.commit();
    expect(await f2.fileType('f.txt')).toBe(FileType.EXECUTABLE);
  });

  it('writeFromFile missing file throws', async () => {
    const b = snap.batch();
    await expect(b.writeFromFile('x.txt', '/nonexistent/file.txt')).rejects.toThrow();
  });

  it('batch mode parameter', async () => {
    const b = snap.batch();
    await b.write('run.sh', toBytes('#!/bin/sh'), { mode: MODE_BLOB_EXEC });
    const f2 = await b.commit();
    expect(await f2.fileType('run.sh')).toBe(FileType.EXECUTABLE);
  });

  it('writeSymlink basic', async () => {
    const b = snap.batch();
    await b.writeSymlink('link.txt', 'a.txt');
    const f2 = await b.commit();
    expect(await f2.readlink('link.txt')).toBe('a.txt');
  });

  it('writeSymlink filemode', async () => {
    const b = snap.batch();
    await b.writeSymlink('link.txt', 'a.txt');
    const f2 = await b.commit();
    expect(await f2.fileType('link.txt')).toBe(FileType.LINK);
  });

  it('writeSymlink with other ops', async () => {
    const b = snap.batch();
    await b.writeSymlink('link.txt', 'a.txt');
    await b.write('new.txt', toBytes('new'));
    const f2 = await b.commit();
    expect(await f2.readlink('link.txt')).toBe('a.txt');
    expect(await f2.exists('new.txt')).toBe(true);
  });

  it('writeSymlink after commit raises', async () => {
    const b = snap.batch();
    await b.commit();
    await expect(b.writeSymlink('link.txt', 'a.txt')).rejects.toThrow(/closed/);
  });

  it('identical writes no new commit', async () => {
    const b = snap.batch();
    await b.write('a.txt', toBytes('a')); // same as existing
    const f2 = await b.commit();
    expect(f2.commitHash).toBe(snap.commitHash);
  });
});

File: kotlin/src/test/kotlin/vost/BatchTest.kt
package vost

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import java.io.FileNotFoundException
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

class BatchTest {

    @Test
    fun `batch write and commit`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.write("a.txt", "a".toByteArray())
            batch.write("b.txt", "b".toByteArray())
            val result = batch.commit()
            assertEquals("a", result.readText("a.txt"))
            assertEquals("b", result.readText("b.txt"))
        }
    }

    @Test
    fun `batch result is set after commit`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            assertNull(batch.result)
            batch.write("file.txt", "data".toByteArray())
            batch.commit()
            assertNotNull(batch.result)
        }
    }

    @Test
    fun `batch auto-close commits`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.write("file.txt", "data".toByteArray())
            batch.close()
            assertNotNull(batch.result)
            assertEquals("data", batch.result!!.readText("file.txt"))
        }
    }

    @Test
    fun `batch use block auto-close`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            var resultFs: Fs? = null
            fs.batch().use { batch ->
                batch.write("file.txt", "data".toByteArray())
                resultFs = null // will be set after close
            }
            // Re-read from the branch
            val newFs = it.branches["main"]
            assertEquals("data", newFs.readText("file.txt"))
        }
    }

    @Test
    fun `batch writeText`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.writeText("msg.txt", "hello")
            val result = batch.commit()
            assertEquals("hello", result.readText("msg.txt"))
        }
    }

    @Test
    fun `batch writeSymlink`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.writeSymlink("link", "target.txt")
            val result = batch.commit()
            assertEquals(FileType.LINK, result.fileType("link"))
            assertEquals("target.txt", result.readlink("link"))
        }
    }

    @Test
    fun `batch remove`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("a.txt", "a".toByteArray())
            fs = fs.write("b.txt", "b".toByteArray())

            val batch = fs.batch()
            batch.remove("a.txt")
            val result = batch.commit()
            assertFalse(result.exists("a.txt"))
            assertTrue(result.exists("b.txt"))
        }
    }

    @Test
    fun `batch remove nonexistent throws`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            assertThrows<FileNotFoundException> {
                batch.remove("nonexistent.txt")
            }
        }
    }

    @Test
    fun `batch remove directory throws`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("dir/file.txt", "data".toByteArray())

            val batch = fs.batch()
            assertThrows<IsADirectoryException> {
                batch.remove("dir")
            }
        }
    }

    @Test
    fun `batch empty commit returns same fs`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            val result = batch.commit()
            assertEquals(fs.commitHash, result.commitHash)
        }
    }

    @Test
    fun `batch closed after commit`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.write("file.txt", "data".toByteArray())
            batch.commit()
            assertThrows<IllegalStateException> {
                batch.write("another.txt", "data".toByteArray())
            }
        }
    }

    @Test
    fun `batch write then remove same path`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "original".toByteArray())

            val batch = fs.batch()
            batch.write("file.txt", "updated".toByteArray())
            batch.remove("file.txt")
            val result = batch.commit()
            assertFalse(result.exists("file.txt"))
        }
    }

    @Test
    fun `batch remove then write same path`() {
        val store = createStore()
        store.use {
            var fs = it.branches["main"]
            fs = fs.write("file.txt", "original".toByteArray())

            val batch = fs.batch()
            batch.remove("file.txt")
            batch.write("file.txt", "new content".toByteArray())
            val result = batch.commit()
            assertEquals("new content", result.readText("file.txt"))
        }
    }

    @Test
    fun `batch with custom message`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch(message = "custom message")
            batch.write("file.txt", "data".toByteArray())
            val result = batch.commit()
            assertEquals("custom message", result.message)
        }
    }

    @Test
    fun `batch on readonly throws`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            it.tags["v1"] = fs
            val tagFs = it.tags["v1"]
            assertThrows<ReadOnlyError> {
                tagFs.batch()
            }
        }
    }

    @Test
    fun `multiple writes to same path keeps last`() {
        val store = createStore()
        store.use {
            val fs = it.branches["main"]
            val batch = fs.batch()
            batch.write("file.txt", "v1".toByteArray())
            batch.write("file.txt", "v2".toByteArray())
            batch.write("file.txt", "v3".toByteArray())
            val result = batch.commit()
            assertEquals("v3", result.readText("file.txt"))
        }
    }
}
